### PR TITLE
140: #140 implement applicative filter DSL

### PR DESCRIPTION
**Pull request by hejfelix**
First stab at implementing an applicative DSL for filters.

### CODE REVIEW

File: modules/benchmarks/FilterBenchmark.scala.output.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  Model Number: Z14Y00047DK/A
+  Chip: Apple M1 Pro
+  Total Number of Cores: 10 (8 performance and 2 efficiency)
+  Memory: 32 GB
+```
+
+```
+[info] # JMH version: 1.32
+[info] # VM version: JDK 17.0.1, OpenJDK 64-Bit Server VM, 17.0.1+12
+[info] # VM invoker: /Users/dkFePaHa/.asdf/installs/java/adoptopenjdk-17.0.1+12/bin/java
```
Comment (Colin-TUE) [inline modules/benchmarks/FilterBenchmark.scala.output.md:26]: Do we want these locations to be in the code?

File: modules/benchmarks/FilterBenchmark.scala.output.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+## Interpretation
+
+The results look similar to what I would expect. The benchmarks log 3 times at different levels (Info, Warn, Error) in a
+tight loop of 1000 iterations. `testEverything` uses `Filter.everything`, i.e. it has to _always_ materialize the log
+output. This is the worst case scenario.
+
+`testInfo` uses the `Filter.exactLevel(LogLevel.Info)` which will only materialize `1/3` of the lines. This is almost 3
+times as fast as `everything`.
+
+`testNothing` is the best case scenario where nothing is materialized. This is more than 3 times faster, since not only
```
Comment (Colin-TUE) [inline modules/benchmarks/FilterBenchmark.scala.output.md:61]: "More than 3 times faster" than what?

File: modules/core/shared/src/main/scala/org/legogroup/woof/Filter.scala [hunk 3]
```diff
@@ -7,22 +7,46 @@ import scala.util.matching.Regex
...
     def empty: Filter                         = nothing
     def combine(f: Filter, g: Filter): Filter = f or g
 
-end Filter
+  extension (f: Filter)
+    infix def and(g: Filter): Filter = Filter.CompositeAnd(f, g)
+    infix def or(g: Filter): Filter  = Filter.CompositeOr(f, g)
+    def interpret: LogLine => Boolean = line =>
+      f match
+        case Filter.AtLeastLevel(level)      => line.level >= level
```
Comment (Colin-TUE) [inline modules/core/shared/src/main/scala/org/legogroup/woof/Filter.scala:41]: A large part of this code looks similar to the code in the `DefaultLogger`.

File: modules/benchmarks/FilterBenchmark.scala.output.md (hunk 1 above)
Comment (hejfelix) [reply modules/benchmarks/FilterBenchmark.scala.output.md:26]: I guess all the neccessary info is in the line above – I'll remove it

File: modules/benchmarks/FilterBenchmark.scala.output.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+[info] Benchmark                        Mode  Cnt      Score     Error  Units
+[info] FilterBenchmark.testEverything  thrpt   25    255,250 ±  12,763  ops/s
+[info] FilterBenchmark.testInfo        thrpt   25    722,664 ±  25,608  ops/s
+[info] FilterBenchmark.testNothing     thrpt   25  11275,189 ± 314,631  ops/s
+[success] Total time: 1509 s (25:09), completed 31 Mar 2023, 10.38.35
+```
+
+## Interpretation
+
+The results look similar to what I would expect. The benchmarks log 3 times at different levels (Info, Warn, Error) in a
```
[the remark this answers is not in the document]
Comment (hejfelix) [reply modules/benchmarks/FilterBenchmark.scala.output.md:54]: will elaborate

**Pull request merged**